src/cast_core/src/SamplingLoader.cpp

```diff
@@ -0,0 +1,137 @@
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cast_core/actors/SamplingLoader.hpp>
+
+#include <algorithm>
+#include <iostream>
+#include <memory>
+
+#include <bsoncxx/builder/basic/document.hpp>
+#include <bsoncxx/builder/basic/kvp.hpp>
+#include <bsoncxx/builder/stream/document.hpp>
+#include <bsoncxx/json.hpp>
+
+#include <mongocxx/client.hpp>
+#include <mongocxx/pool.hpp>
+
+#include <boost/log/trivial.hpp>
+
+#include <gennylib/Cast.hpp>
+#include <gennylib/MongoException.hpp>
+#include <gennylib/context.hpp>
+
+#include <value_generators/DocumentGenerator.hpp>
+#include <value_generators/Pipeline.hpp>
+
+namespace genny::actor {
+
+/** @private */
+using index_type = std::tuple<DocumentGenerator, std::optional<DocumentGenerator>, std::string>;
+using namespace bsoncxx;
+
+using bsoncxx::builder::basic::kvp;
+using bsoncxx::builder::basic::make_document;
+
+/** @private */
+struct SamplingLoader::PhaseConfig {
+    PhaseConfig(PhaseContext& context, mongocxx::pool::entry& client, ActorId id)
+        : database{(*client)[context["Database"].to<std::string>()]},
+          collection{database[context["Collection"].to<std::string>()]},
+          // The next line uses integer division for non multithreaded configurations.
+          // The Remainder is accounted for below.
+          insertBatchSize{context["InsertBatchSize"].to<IntegerSpec>()},
+          numBatches{context["Batches"].to<IntegerSpec>()},
+          sampleSize{
+              context["SampleSize"].maybe<IntegerSpec>().value_or(numBatches * insertBatchSize)},
+          pipelineSuffix{context["Pipeline"].maybe<Pipeline>(context, id).value_or(Pipeline{})} {}
+
+    mongocxx::database database;
+    mongocxx::collection collection;
+    int64_t insertBatchSize;
+    int64_t numBatches;
+    int64_t sampleSize;
+    Pipeline pipelineSuffix;
+};
+
+void genny::actor::SamplingLoader::run() {
+    for (auto&& config : _loop) {
+        for (auto&& _ : config) {
+            BOOST_LOG_TRIVIAL(debug) << "Beginning to run SamplingLoader";
+            // Read the sample size documents into a vector.
+            mongocxx::pipeline samplePipeline;
+            samplePipeline.sample(config->sampleSize);
+            samplePipeline.project(make_document(kvp("_id", 0)));
+            samplePipeline.append_stages(config->pipelineSuffix.generatePipeline().view_array());
+            auto cursor =
+                config->collection.aggregate(samplePipeline, mongocxx::options::aggregate{});
+            const auto sampleDocs =
+                std::vector<bsoncxx::document::value>(cursor.begin(), cursor.end());
+
+            if (sampleDocs.empty()) {
```

Comment:
The `sampleDocs` might end up empty if the workload adds filtering into the pipeline and, while I cannot think of any real workload that would want to do it, I think we should at least document it for the phase config, because the failure might depend on the sampled data and would be hard to diagnose on Evergreen due to not being able to reproduce the sample.